utils/communication_report.py

```diff
@@ -0,0 +1,133 @@
+    except json.JSONDecodeError:
+        raise Exception(f"get_fhir_resource({url}), failed to parse JSON from command output")
+
+#    response = requests.get(url)
+#    response.raise_for_status()  # Ensure we stop on HTTP errors
+#    return response.json()
+
+def construct_next_url(next_url):
+    # Parse the query parameters from the next URL
+    next_query_params = urllib.parse.urlparse(next_url).query
+    #print(f"construct_next_url({next_url}), here's next_query_params:{next_query_params}")
+    next_params = urllib.parse.parse_qs(next_query_params)
+
+    # Extract needed parameters
+    getpages = next_params.get('_getpages', [''])[0]
+    getpagesoffset = next_params.get('_getpagesoffset', [''])[0]
+    count = next_params.get('_count', [''])[0]
+    bundletype = next_params.get('_bundletype', [''])[0]
+
+    # Construct the new URL
+    new_url = f"{base_url}\?_getpages={getpages}\&_getpagesoffset={getpagesoffset}\&_count={count}\&_bundletype={bundletype}"
+    return new_url
+
+def get_patient_isacc_id(patient_reference, patient_cache):
+    # Check if we already have the ISACC ID for this patient
+    if patient_reference in patient_cache:
+        return patient_cache[patient_reference]
+
+    patient_url = f"{base_url}/{patient_reference}"
+    patient_data = get_fhir_resource(patient_url)
+
+    # Initialize ISACC ID as empty string
+    isacc_id = ""
+
+    # Search for ISACC ID in the patient's identifiers
+    for identifier in patient_data.get('identifier', []):
+        if identifier.get('system') == "http://isacc.app/user-id":
+            isacc_id = identifier.get('value', "")
+            break
+
+    # Cache the ISACC ID for future use
+    patient_cache[patient_reference] = isacc_id
+    return isacc_id
+
+def extract_type(communication):
+    for category in communication.get('category', []):
+        for coding in category.get('coding', []):
+            if coding.get('system') == "https://isacc.app/CodeSystem/communication-type":
+                return coding.get('code')
+    return None
+
+def main():
+    url = f"{base_url}/Communication?_sort=sent"
+    patient_cache = {}
+
+    with open('CommunicationReport.csv', 'w', newline='') as file:
+        writer = csv.writer(file)
+        writer.writerow(["ID", "Datetime", "Recipient FHIR ID", "Recipient ISACC ID", "Type", "Content", "Sender", "Note"])
+
+        while url:
+            data = get_fhir_resource(url)
+            for entry in data.get('entry', []):
+                communication = entry['resource']
+                communication_id = communication.get('id')
+                datetime = communication.get('sent')
+                type = extract_type(communication)
+                if type in ['isacc-comment', 'isacc-non-sms-message']:
+                    recipient_reference = communication.get('subject', {}).get('reference')
+                else:
+                    recipient_reference = communication.get('recipient', [{}])[0].get('reference')
```

Comment:
jsonpath (`jmespath`) expressions could simplify some of these expressions, not 100% sure how it'd work with missing values though...

[here's an example from cosri](https://github.com/uwcirg/cosri-patientsearch/blob/1fd570c4e8a0d4398db0d417aa5cc1feb7d30363/tests/test_columnconfig.py#L22)

Reply:
Thanks for the pointer - funny, I just searched the cosri codebases today for jsonpath and didn't find a reference. I'm wrapping up another script to analyze potential duplicate UW prod cosri Patient resources, might use `jmespath` there.

Reply:
The python library support for JSONPath hasn't been great; JMESPath is a similar standard, but had a better python library